File: Application/Frontend/website/src/features/partners/services/partnerService.js
import { collection, addDoc, query, where, getDocs, doc, onSnapshot, serverTimestamp } from 'firebase/firestore';
import { db } from '../../../firebase';
import { calculateSocietyMonthlyFee } from '../../../utils/pricingEngine';

/**
 * GateLink Partner & Lead Firestore Service
 */
export const submitPartnerLead = async (leadData) => {
  const generatedRef = `LEAD-${Date.now().toString().slice(-6)}`;
  
  const docRef = await addDoc(collection(db, 'partner_leads'), {
    referenceId: generatedRef,
    partnerName: leadData.partnerName.trim(),
    partnerPhone: leadData.partnerPhone.trim(),
    partnerEmail: leadData.partnerEmail.trim(),
    partnerCity: leadData.partnerCity.trim(),
    partnerUpi: leadData.partnerUpi.trim(),
    partnerType: leadData.partnerType || 'broker',
    referredByCode: leadData.referredByCode?.trim() || null,
    targetSocietyName: leadData.targetSocietyName.trim(),
    targetCity: leadData.targetCity.trim(),
    contactPerson: leadData.contactPerson.trim(),
    contactRole: leadData.contactRole || 'RWA Secretary',
    contactPhone: leadData.contactPhone.trim(),
    approxFlats: leadData.approxFlats || '100-250',
    notes: leadData.notes?.trim() || '',
    status: 'new',
    assignedTier: leadData.assignedTier || 'growth',
    source: 'website_partners_portal',
    createdAt: serverTimestamp(),
  });

  return { id: docRef.id, referenceId: generatedRef };
};

/**
 * Fetch all leads submitted by a partner's phone number or Reference ID
 */
export const lookupPartnerLeads = async (searchQuery) => {
  const cleanQuery = searchQuery.trim();
  const leadsRef = collection(db, 'partner_leads');
  
  // Try querying by phone first
  const qPhone = query(leadsRef, where('partnerPhone', '==', cleanQuery));
  const phoneSnap = await getDocs(qPhone);
  
  if (!phoneSnap.empty) {
    return phoneSnap.docs.map((doc) => ({ id: doc.id, ...doc.data() }));
  }

  // Fallback: Query by referenceId (e.g. LEAD-123456)
  const qRef = query(leadsRef, where('referenceId', '==', cleanQuery.toUpperCase()));
  const refSnap = await getDocs(qRef);

  return refSnap.docs.map((doc) => ({ id: doc.id, ...doc.data() }));
};

/**
 * Real-time listener for dynamic partner commission percentage rates
 */
export const subscribeCommissionRates = (onUpdate) => {
  const configDocRef = doc(db, 'system_config', 'partner_program');
  return onSnapshot(
    configDocRef,
    (snap) => {
      if (snap.exists()) {
        onUpdate(snap.data());
      }
    },
    (err) => {
      console.warn('Notice: system_config/partner_program using default rates:', err);
    }
  );
};

/**
 * Partner Self-Serve Direct Society Onboarding & Secret Activation Link Generator
 */
export const onboardSocietyByPartner = async (onboardingData) => {
  const generatedRef = `LEAD-${Date.now().toString().slice(-6)}`;
  const activationCode = `SOC-${Math.floor(100000 + Math.random() * 900000)}`;

  const flatCountNum = Number(onboardingData.flatCount) || 100;
  const calculatedMrr = calculateSocietyMonthlyFee(flatCountNum);

  // 1. Save Partner Lead & Onboarding Record
  const leadRef = await addDoc(collection(db, 'partner_leads'), {
    referenceId: generatedRef,
    activationCode,
    isOnboardedByPartner: true,
    partnerName: onboardingData.partnerName.trim(),
    partnerPhone: onboardingData.partnerPhone.trim(),
    partnerEmail: onboardingData.partnerEmail?.trim() || '',
    partnerUpi: onboardingData.partnerUpi.trim(),
    partnerType: onboardingData.partnerType || 'onboarding_partner',

    targetSocietyName: onboardingData.societyName.trim(),
    targetCity: onboardingData.city.trim(),
    address: onboardingData.address.trim(),
    pincode: onboardingData.pincode?.trim() || '',
    
    wings: onboardingData.wings || [],
    approxFlats: String(flatCountNum),
    flatCount: flatCountNum,
    gatesCount: Number(onboardingData.gatesCount) || 2,
    guardDevicesCount: Number(onboardingData.guardDevicesCount) || 2,
    mrr: calculatedMrr,
    monthlyFee: calculatedMrr,

    contactPerson: onboardingData.rwaSecretaryName.trim(),
    contactPhone: onboardingData.rwaSecretaryPhone.trim(),
    contactEmail: onboardingData.rwaSecretaryEmail?.trim() || '',
    contactRole: 'RWA Secretary',

    status: 'won', // Self-onboarded by partner
    assignedTier: onboardingData.assignedTier || 'onboarding',
    source: 'partner_self_serve_onboarding_wizard',
    createdAt: serverTimestamp(),
  });

  // 2. Create Pending Provisioned Society Record
  await addDoc(collection(db, 'societies'), {
    name: onboardingData.societyName.trim(),
    city: onboardingData.city.trim(),
    address: onboardingData.address.trim(),
    code: activationCode,
    activationCode,
    partnerReferenceId: generatedRef,
    partnerLeadId: leadRef.id,
    flatCount: flatCountNum,
    wings: onboardingData.wings || [],
    gatesCount: Number(onboardingData.gatesCount) || 2,
    president: onboardingData.rwaSecretaryName.trim(),
    phone: onboardingData.rwaSecretaryPhone.trim(),
    contactPerson: onboardingData.rwaSecretaryName.trim(),
    contactPhone: onboardingData.rwaSecretaryPhone.trim(),
    mrr: calculatedMrr,
    monthlyFee: calculatedMrr,
    status: 'pending_rwa_activation',
    createdAt: serverTimestamp(),
  });

  return {
    id: leadRef.id,
    referenceId: generatedRef,
    activationCode,
    activationUrl: `https://app.gatelink.in/activate?code=${activationCode}&ref=${generatedRef}`,
    rwaPhone: onboardingData.rwaSecretaryPhone.trim(),
    rwaName: onboardingData.rwaSecretaryName.trim(),
    societyName: onboardingData.societyName.trim(),
  };
};

File: Application/Frontend/partner_app/lib/features/leads/data/partner_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';

class PartnerLeadModel {
  final String id;
  final String referenceId;
  final String targetSocietyName;
  final String targetCity;
  final String contactPerson;
  final String contactPhone;
  final String approxFlats;
  final String status; // new, contacted, demo_scheduled, won, lost
  final String payoutStatus; // pending, processing, paid
  final double payoutTotal;
  final String utrNumber;
  final double monthlyPassiveEarned;
  final String partnerName;
  final String partnerPhone;
  final DateTime? createdAt;
  final DateTime? paidAt;

  PartnerLeadModel({
    required this.id,
    required this.referenceId,
    required this.targetSocietyName,
    required this.targetCity,
    required this.contactPerson,
    required this.contactPhone,
    required this.approxFlats,
    required this.status,
    required this.payoutStatus,
    required this.payoutTotal,
    required this.utrNumber,
    required this.monthlyPassiveEarned,
    required this.partnerName,
    required this.partnerPhone,
    this.createdAt,
    this.paidAt,
  });

  factory PartnerLeadModel.fromFirestore(DocumentSnapshot doc) {
    final data = doc.data() as Map<String, dynamic>? ?? {};
    return PartnerLeadModel(
      id: doc.id,
      referenceId: data['referenceId'] ?? 'LEAD-000',
      targetSocietyName: data['targetSocietyName'] ?? 'Society',
      targetCity: data['targetCity'] ?? 'City',
      contactPerson: data['contactPerson'] ?? 'Secretary',
      contactPhone: data['contactPhone'] ?? '',
      approxFlats: data['approxFlats']?.toString() ?? '100',
      status: data['status'] ?? 'new',
      payoutStatus: data['payoutStatus'] ?? 'pending',
      payoutTotal: (data['payoutTotal'] ?? 0).toDouble(),
      utrNumber: data['utrNumber'] ?? '',
      monthlyPassiveEarned: (data['monthlyPassiveEarned'] ?? 0).toDouble(),
      partnerName: data['partnerName'] ?? '',
      partnerPhone: data['partnerPhone'] ?? '',
      createdAt: (data['createdAt'] as Timestamp?)?.toDate(),
      paidAt: (data['paidAt'] as Timestamp?)?.toDate(),
    );
  }

  Map<String, dynamic> toMap() => {
    'referenceId': referenceId,
    'targetSocietyName': targetSocietyName,
    'targetCity': targetCity,
    'contactPerson': contactPerson,
    'contactPhone': contactPhone,
    'approxFlats': approxFlats,
    'status': status,
    'payoutStatus': payoutStatus,
    'payoutTotal': payoutTotal,
    'utrNumber': utrNumber,
    'monthlyPassiveEarned': monthlyPassiveEarned,
    'partnerName': partnerName,
    'partnerPhone': partnerPhone,
    'createdAt': FieldValue.serverTimestamp(),
  };
}

class PartnerService {
  final FirebaseFirestore _db = FirebaseFirestore.instance;

  // Stream partner leads live by phone number
  Stream<List<PartnerLeadModel>> streamPartnerLeads(String partnerPhone) {
    final phoneClean = partnerPhone.replaceAll(RegExp(r'[^0-9]'), '');

    return _db
        .collection('partner_leads')
        .where('partnerPhone', isEqualTo: phoneClean)
        .snapshots()
        .map((snapshot) => snapshot.docs
            .map((doc) => PartnerLeadModel.fromFirestore(doc))
            .toList());
  }

  // Submit new lead to Firestore
  Future<void> submitLead({
    required String targetSocietyName,
    required String targetCity,
    required String contactPerson,
    required String contactPhone,
    required String approxFlats,
    required String partnerName,
    required String partnerPhone,
  }) async {
    final phoneClean = partnerPhone.replaceAll(RegExp(r'[^0-9]'), '');
    final docId = 'LEAD-${DateTime.now().millisecondsSinceEpoch}';

    await _db.collection('partner_leads').doc(docId).set({
      'referenceId': 'LEAD-${DateTime.now().millisecondsSinceEpoch.toString().substring(6)}',
      'targetSocietyName': targetSocietyName,
      'targetCity': targetCity,
      'contactPerson': contactPerson,
      'contactPhone': contactPhone,
      'approxFlats': approxFlats,
      'status': 'new',
      'payoutStatus': 'pending',
      'payoutTotal': 500,
      'utrNumber': '',
      'monthlyPassiveEarned': 0,
      'partnerName': partnerName,
      'partnerPhone': phoneClean,
      'createdAt': FieldValue.serverTimestamp(),
    });
  }
}
